File: python/2962.count_subarrays_where_max_element_appears_at_least_k_times.py
from typing import List


class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:

        max_num = max(nums)
        left = 0
        result = 0
        count = 0
        n = len(nums)

        for right in range(n):
            if nums[right] == max_num:
                count += 1
            while count >= k:
                result += n - right
                if nums[left] == max_num:
                    count -= 1
                left += 1
        return result

File: cpp/2962.count_subarrays_where_max_element_appears_at_least_k_times.cpp
#include <cstdint>
#include <vector>

class Solution {
 public:
    std::int64_t countSubarrays(const std::vector<int>& nums, int k) {
        std::int64_t max_num =
            *std::max_element(std::begin(nums), std::end(nums));
        std::int64_t result = 0;
        auto left = 0;
        auto count = 0;

        auto right = 0;
        for (auto num : nums) {
            if (nums[right] == max_num) {
                ++count;
            }
            while (count >= k) {
                result += (nums.size() - right);
                if (nums[left] == max_num) {
                    --count;
                }
                ++left;
            }
            ++right;
        }
        return result;
    }
};
